# EDA PRELIMINAR

In [1]:
#tratamiento de datos
import pandas as pd

#configuracion de visualizacion
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows', None)

#exploracion de archivos
import sys
sys.path.append('../')

#funciones personalizadas
from SRC import sp_eda as se

In [2]:
df_u = pd.read_csv('../datos/olist_unificado.csv')

### Eda_rapido identificación de columnas irrelevantes

In [3]:
se.eda_rapido(df_u)

MUESTRA:


order_id                       customer_id  \
35352  f60e24afa1a33daf8c332325bafb1687  bf6313d9148432b6ac91c824e263e86a   
16420  3304c0c857a9c77a201a551f5a3cacb8  5a47dac0d86df8b85578ff714c28b4f3   
36610  d2ef78d642086d0cea1193953c8732be  6d54acc72b9bb27db5e46f04fe577e74   
8864   16656bc6faba3fd5740dd2305e5a9a94  ce88f748efb459730857f1a6513d1ae5   
11034  ef5025f48c69b9d957744cd6d6f9b662  1cd53452b9fbe5e67eb6d89448ed09f9   

      order_status order_purchase_timestamp    order_approved_at  \
35352    delivered      2018-03-04 17:48:42  2018-03-05 18:31:21   
16420    delivered      2017-05-04 12:57:21  2017-05-04 13:10:46   
36610    delivered      2017-11-24 19:46:05  2017-11-25 05:16:33   
8864     delivered      2017-07-14 19:23:44  2017-07-15 03:35:35   
11034    delivered      2017-11-24 14:41:23  2017-11-24 16:52:55   

      order_delivered_carrier_date order_delivered_customer_date  \
35352          2018-03-10 02:41:32           2018-03-21 21:04:01   
16420          2017-05-05 11:50:12           2017-05-08 10:27:31   
36610          2017-11-29 13:29:13           2017-12-04 20:42:32   
8864           2017-07-17 19:13:03           2017-07-24 18:13:39   
11034          2017-11-30 16:39:02           2017-12-19 12:03:22   

      order_estimated_delivery_date  order_item_id  \
35352           2018-03-22 00:00:00            1.0   
16420           2017-05-25 00:00:00            3.0   
36610           2017-12-18 00:00:00            1.0   
8864            2017-08-09 00:00:00            1.0   
11034           2017-12-22 00:00:00            1.0   

                             product_id                         seller_id  \
35352  3b0f7951038b105522c2d566b54421f7  ea8482cd71df3c1969d7b9473ff13abc   
16420  7bbc646134de228a74392d8407d16df8  ddd51ae8cda92f3995a51fc0f0f3eec7   
36610  0d86ff0b639e969a11a7b3dce67e181c  3b15288545f8928d3e65a8f949a28291   
8864   be837f2e0152a208d4386f4126d5bd7c  f8db351d8c4c4c22c6835c19a46f01b0   
11034  57d79905de06d8897872c551bfd09358  ea8482cd71df3c1969d7b9473ff13abc   

       shipping_limit_date  price  freight_value  payment_sequential  \
35352  2018-03-09 17:55:23  29.98          15.10                 1.0   
16420  2017-05-10 13:10:46  34.90           9.34                 1.0   
36610  2017-12-04 05:16:33  89.99          13.65                 1.0   
8864   2017-07-20 03:34:09  32.90          17.60                 1.0   
11034  2017-11-30 16:52:55  21.99          15.10                 1.0   

      payment_type  payment_installments  payment_value  \
35352  credit_card                   1.0          45.08   
16420   debit_card                   1.0         265.44   
36610       boleto                   1.0         207.28   
8864        boleto                   1.0          50.50   
11034  credit_card                   1.0          37.09   

                              review_id  review_score review_comment_title  \
35352  d1e0dd5e52c0125d01dbd65508d127a3           5.0                  NaN   
16420  5950f117c703b67b5935506f4e5c79f1           5.0                  NaN   
36610  08ebd29e891585c2f5831c6fd8b052f6           5.0                  NaN   
8864   fd4e5ff69f8cb5665852631035e0c348           4.0                  NaN   
11034  de1ebe4fdbca762e522ff43f1ca66ace           4.0                  NaN   

      review_comment_message review_creation_date review_answer_timestamp  \
35352                    NaN  2018-03-22 00:00:00     2018-03-23 17:53:39   
16420                    NaN  2017-05-09 00:00:00     2017-05-18 15:59:30   
36610                    NaN  2017-12-05 00:00:00     2017-12-06 09:50:13   
8864                     NaN  2017-07-25 00:00:00     2017-07-26 03:13:20   
11034                    NaN  2017-12-20 00:00:00     2017-12-20 16:14:29   

                     customer_unique_id  customer_zip_code_prefix  \
35352  38a90d5218eb2a167f266465cd550168                     37410   
16420  cd336c2a4679b726894657837bf1dca7                     21031   
36610  b83736ca9d8ee95188aebbe3e590

------------------------
INFO:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119143 entries, 0 to 119142
Data columns (total 40 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       119143 non-null  object 
 1   customer_id                    119143 non-null  object 
 2   order_status                   119143 non-null  object 
 3   order_purchase_timestamp       119143 non-null  object 
 4   order_approved_at              118966 non-null  object 
 5   order_delivered_carrier_date   117057 non-null  object 
 6   order_delivered_customer_date  115722 non-null  object 
 7   order_estimated_delivery_date  119143 non-null  object 
 8   order_item_id                  118310 non-null  float64
 9   product_id                     118310 non-null  object 
 10  seller_id                      118310 non-null  object 
 11  shipping_limit_date            118310 non-null  object 
 12 

None

------------------------
NULOS:


order_id                          0.000000
customer_id                       0.000000
order_status                      0.000000
order_purchase_timestamp          0.000000
order_approved_at                 0.148561
order_delivered_carrier_date      1.750837
order_delivered_customer_date     2.871339
order_estimated_delivery_date     0.000000
order_item_id                     0.699160
product_id                        0.699160
seller_id                         0.699160
shipping_limit_date               0.699160
price                             0.699160
freight_value                     0.699160
payment_sequential                0.002518
payment_type                      0.002518
payment_installments              0.002518
payment_value                     0.002518
review_id                         0.836810
review_score                      0.836810
review_comment_title             88.258647
review_comment_message           57.827988
review_creation_date              0.836810
review_answ

------------------------
ESTADISTICOS NUMERICAS:


count          mean           std      min  \
order_item_id               118310.0      1.196543      0.699489     1.00   
price                       118310.0    120.646603    184.109691     0.85   
freight_value               118310.0     20.032387     15.836850     0.00   
payment_sequential          119140.0      1.094737      0.730141     1.00   
payment_installments        119140.0      2.941246      2.777848     0.00   
payment_value               119140.0    172.735135    267.776077     0.00   
review_score                118146.0      4.015582      1.400436     1.00   
customer_zip_code_prefix    119143.0  35033.451298  29823.198969  1003.00   
product_name_lenght         116601.0     48.767498     10.033540     5.00   
product_description_lenght  116601.0    785.967822    652.584121     4.00   
product_photos_qty          116601.0      2.205161      1.717452     1.00   
product_weight_g            118290.0   2112.250740   3786.695111     0.00   
product_length_cm           118290.0     30.265145     16.189367     7.00   
product_height_cm           118290.0     16.619706     13.453584     2.00   
product_width_cm            118290.0     23.074799     11.749139     6.00   
seller_zip_code_prefix      118310.0  24442.410413  27573.004511  1001.00   

                                 25%       50%       75%       max  
order_item_id                   1.00      1.00      1.00     21.00  
price                          39.90     74.90    134.90   6735.00  
freight_value                  13.08     16.28     21.18    409.68  
payment_sequential              1.00      1.00      1.00     29.00  
payment_installments            1.00      2.00      4.00     24.00  
payment_value                  60.85    108.16    189.24  13664.08  
review_score                    4.00      5.00      5.00      5.00  
customer_zip_code_prefix    11250.00  24240.00  58475.00  99990.00  
product_name_lenght            42.00     52.00     57.00     76.00  
product_description_lenght    346.00    600.00    983.00   3992.00  
product_photos_qty              1.00      1.00      3.00     20.00  
product_weight_g              300.00    700.00   1800.00  40425.00  
product_length_cm              18.00     25.00     38.00    105.00  
product_height_cm               8.00     13.00     20.00    105.00  
product_width_cm               15.00     20.00     30.00    118.00  
seller_zip_code_prefix       6429.00  13660.00  27972.00  99730.00

### Eliminación de columnas irrelevantes

In [4]:
columnas_irrelevantes= [
    'review_comment_title',
    'review_comment_message', 
    'customer_unique_id',
    'review_answer_timestamp',
    'payment_sequential',
    'order_item_id',
    'shipping_limit_date',
    'product_length_cm',
    'product_height_cm',
    'product_width_cm',
    'product_category_name',
    'customer_city',
    'seller_city'
]

df_u = df_u.drop(columns=columnas_irrelevantes)


### Reuvicacion de columnas ID al final del data frame para facilitar visualizacion

In [5]:
id_cols = ['order_id','customer_id','product_id','seller_id','review_id']
cols = [c for c in df_u.columns if c not in id_cols] + id_cols
df_u = df_u[cols]

### EDA PRELIMINAR

In [6]:
se.eda_preliminar(df_u)

order_status order_purchase_timestamp    order_approved_at  \
99446     delivered      2018-01-02 14:13:47  2018-01-03 14:07:25   
75216     delivered      2017-08-25 21:45:09  2017-08-26 02:11:17   
4732      delivered      2017-09-01 17:06:54  2017-09-05 04:30:47   
108084    delivered      2017-05-24 19:03:52  2017-05-24 19:15:23   
66633     delivered      2018-03-26 16:47:13  2018-03-26 16:55:36   

       order_delivered_carrier_date order_delivered_customer_date  \
99446           2018-01-13 00:43:05           2018-01-16 18:33:41   
75216           2017-08-28 19:23:01           2017-08-30 19:27:48   
4732            2017-09-06 19:54:56           2017-09-13 20:53:43   
108084          2017-05-25 17:08:15           2017-06-01 12:30:03   
66633           2018-04-03 13:03:42           2018-04-19 16:38:37   

       order_estimated_delivery_date   price  freight_value payment_type  \
99446            2018-01-18 00:00:00   59.00           8.72  credit_card   
75216            2017-09-06 00:00:00   24.99           7.78       boleto   
4732             2017-09-20 00:00:00   39.90          15.10       boleto   
108084           2017-06-16 00:00:00  129.90          34.27  credit_card   
66633            2018-04-25 00:00:00   61.99          37.12  credit_card   

        payment_installments  payment_value  review_score  \
99446                    1.0          67.72           5.0   
75216                    1.0          32.77           3.0   
4732                     1.0          55.00           4.0   
108084                   4.0         164.17           4.0   
66633                    3.0          99.11           5.0   

       review_creation_date  customer_zip_code_prefix customer_state  \
99446   2018-01-17 00:00:00                      4660             SP   
75216   2017-08-31 00:00:00                      8072             SP   
4732    2017-09-14 00:00:00                     11020             SP   
108084  2017-06-02 00:00:00                     84560             PR   
66633   2018-04-20 00:00:00                     59196             RN   

        product_name_lenght  product_description_lenght  product_photos_qty  \
99446                  60.0                       234.0                 1.0   
75216                  51.0                       536.0                 1.0   
4732                   37.0                       265.0                 2.0   
108084                 54.0                       253.0                 4.0   
66633                  50.0                      1257.0                 1.0   

        product_weight_g  seller_zip_code_prefix seller_state  \
99446              600.0                  9720.0           SP   
75216              100.0                  4160.0           SP   
4732               200.0                 31030.0           MG   
108084            6300.0                 17510.0           SP   
66633              350.0                 65072.0           MA   

       product_category_name_english                          order_id  \
99446            musical_instruments  6a9ed89907267cac71cbbf67931313a2   
75216                      telephony  35e43ade963a5aeb52cb24bb5df33ddc   
4732                  consoles_games  82919647c77d6bdf1f3edba9eab8e1fe   
108084                sports_leisure  2ecf490b77cadc218f299cd6678277c8   
66633                  health_beauty  69d68d6cabed593e526a561a94d76d23   

                             customer_id                        product_id  \
99446   827a5d5a014607756a15b47b7d21da37  44e086c4a977f37a888627b43880586c   
75216   22ce9cf0c503b262c657de58e910feff  080a388eb2bd1051b01f6fc4ca659450   
4732    5e3e1f4586f475c893bb9b464e0f198d  4e45c321347cc3336c76fd6350da3445   
108084  a4591c740887b554a3398c3e399cb4f0  0ef97c3cdccb448931cb0f778af3584c   
66633   0a3e4f7efaf1befc4a84e1adcd9b70f5  3fbc0ef745950c7932d5f2a446189725   

                               seller_id                         review_id  
99446   f5a590cf36251cf1162ea35bef76fe84  da41a2f7648bc447d3d91

------------------------
DIMENSIONES
Nuestro conjunto de datos presenta un total de 119143 filas y 27 columnas
------------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119143 entries, 0 to 119142
Data columns (total 27 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_status                   119143 non-null  object 
 1   order_purchase_timestamp       119143 non-null  object 
 2   order_approved_at              118966 non-null  object 
 3   order_delivered_carrier_date   117057 non-null  object 
 4   order_delivered_customer_date  115722 non-null  object 
 5   order_estimated_delivery_date  119143 non-null  object 
 6   price                          118310 non-null  float64
 7   freight_value                  118310 non-null  float64
 8   payment_type                   119140 non-null  object 
 9   payment_installments           119140 non-null  float64
 10  payment_val

None

------------------------
NULOS


order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.148561
order_delivered_carrier_date     1.750837
order_delivered_customer_date    2.871339
order_estimated_delivery_date    0.000000
price                            0.699160
freight_value                    0.699160
payment_type                     0.002518
payment_installments             0.002518
payment_value                    0.002518
review_score                     0.836810
review_creation_date             0.836810
customer_zip_code_prefix         0.000000
customer_state                   0.000000
product_name_lenght              2.133571
product_description_lenght       2.133571
product_photos_qty               2.133571
product_weight_g                 0.715946
seller_zip_code_prefix           0.699160
seller_state                     0.699160
product_category_name_english    2.154554
order_id                         0.000000
customer_id                      0

------------------------
DUPLICADOS
11303
------------------------
FRECUENCIA CATEGORICAS
ORDER_STATUS
order_status
delivered      115723
shipped          1256
canceled          750
unavailable       652
invoiced          378
processing        376
created             5
approved            3
Name: count, dtype: int64
------------------------
PAYMENT_TYPE
payment_type
credit_card    87776
boleto         23190
voucher         6465
debit_card      1706
not_defined        3
Name: count, dtype: int64
------------------------
CUSTOMER_STATE
customer_state
SP    50265
RJ    15518
MG    13819
RS     6573
PR     6043
SC     4345
BA     4091
DF     2516
GO     2466
ES     2360
PE     1906
CE     1565
MT     1132
PA     1129
MS      862
MA      856
PB      644
PI      576
RN      574
AL      464
SE      403
TO      340
RO      292
AM      173
AC       95
AP       84
RR       52
Name: count, dtype: int64
------------------------
SELLER_STATE
seller_state
SP    84377
MG     9314
PR     9096
RJ     5

count          mean           std      min  \
price                       118310.0    120.646603    184.109691     0.85   
freight_value               118310.0     20.032387     15.836850     0.00   
payment_installments        119140.0      2.941246      2.777848     0.00   
payment_value               119140.0    172.735135    267.776077     0.00   
review_score                118146.0      4.015582      1.400436     1.00   
customer_zip_code_prefix    119143.0  35033.451298  29823.198969  1003.00   
product_name_lenght         116601.0     48.767498     10.033540     5.00   
product_description_lenght  116601.0    785.967822    652.584121     4.00   
product_photos_qty          116601.0      2.205161      1.717452     1.00   
product_weight_g            118290.0   2112.250740   3786.695111     0.00   
seller_zip_code_prefix      118310.0  24442.410413  27573.004511  1001.00   

                                 25%       50%       75%       max  
price                          39.90     74.90    134.90   6735.00  
freight_value                  13.08     16.28     21.18    409.68  
payment_installments            1.00      2.00      4.00     24.00  
payment_value                  60.85    108.16    189.24  13664.08  
review_score                    4.00      5.00      5.00      5.00  
customer_zip_code_prefix    11250.00  24240.00  58475.00  99990.00  
product_name_lenght            42.00     52.00     57.00     76.00  
product_description_lenght    346.00    600.00    983.00   3992.00  
product_photos_qty              1.00      1.00      3.00     20.00  
product_weight_g              300.00    700.00   1800.00  40425.00  
seller_zip_code_prefix       6429.00  13660.00  27972.00  99730.00

In [7]:
df_u[df_u.duplicated()].head()

order_status order_purchase_timestamp    order_approved_at  \
14    delivered      2017-05-16 19:41:10  2017-05-16 19:50:18   
60    delivered      2018-07-26 14:39:59  2018-07-26 14:55:10   
66    delivered      2018-03-09 14:50:15  2018-03-09 15:40:39   
76    delivered      2018-01-07 21:44:54  2018-01-07 21:51:28   
79    delivered      2018-06-18 16:02:23  2018-06-18 17:00:57   

   order_delivered_carrier_date order_delivered_customer_date  \
14          2017-05-18 11:40:40           2017-05-29 11:18:31   
60          2018-07-27 12:04:00           2018-07-31 22:26:55   
66          2018-03-15 00:31:19           2018-04-03 13:28:46   
76          2018-01-10 21:56:09           2018-01-17 20:26:31   
79          2018-06-19 15:55:00           2018-06-22 21:18:51   

   order_estimated_delivery_date   price  freight_value payment_type  \
14           2017-06-07 00:00:00   99.00          30.53  credit_card   
60           2018-08-16 00:00:00   60.00          15.52  credit_card   
66           2018-04-02 00:00:00   22.99          22.85  credit_card   
76           2018-02-14 00:00:00  329.90          82.48  credit_card   
79           2018-07-13 00:00:00   84.90          13.25  credit_card   

    payment_installments  payment_value  review_score review_creation_date  \
14                   1.0         259.06           1.0  2017-05-30 00:00:00   
60                   2.0         151.04           1.0  2018-08-01 00:00:00   
66                   9.0          91.68           1.0  2018-04-04 00:00:00   
76                   4.0         824.76           3.0  2018-01-18 00:00:00   
79                   3.0         392.60           5.0  2018-06-23 00:00:00   

    customer_zip_code_prefix customer_state  product_name_lenght  \
14                     20780             RJ                 36.0   
60                     90620             RS                 60.0   
66                     49030             SE                 59.0   
76                     46740             BA                 57.0   
79                      8220             SP                 60.0   

    product_description_lenght  product_photos_qty  product_weight_g  \
14                       450.0                 1.0            9000.0   
60                      1153.0                 2.0             100.0   
66                       566.0                 1.0             150.0   
76                       424.0                 1.0           10000.0   
79                       380.0                 1.0             150.0   

    seller_zip_code_prefix seller_state    product_category_name_english  \
14                 29156.0           ES                     garden_tools   
60                  9628.0           SP                   sports_leisure   
66                  3126.0           SP                        telephony   
76                 80215.0           PR  construction_tools_construction   
79                 17056.0           SP                             auto   

                            order_id                       customer_id  \
14  e6ce16cb79ec1d90b1da9085a6118aeb  494dded5b201313c64ed7f100595b95c   
60  9faeb9b2746b9d7526aef5acb08e2aa0  79183cd650e2bb0d475b0067d45946ac   
66  66e4624ae69e7dc89bd50222b59f581f  684fa6da5134b9e4dab731e00011712d   
76  f271576bed568e896f99eb710cd3a6f8  5dda11942d4f77bee3a46d71e442aec4   
79  68e48e68da1f50f7c5838ea75e3a20dd  4afc1dcca5fe8926fc97d60a4497f8ab   

                          product_id                         seller_id  \
14  08574b074924071f4e201e151b152b4e  001cca7ae9ae17fb1caed9dfb1094831   
60  f48eb5c2fde13ca63664f0bb05f55346  f7ba60f8c3f99e7ee4042fdef03b70c4   
66  b37b72d5a56f887725c2862184b8cab8  db4350fd57ae30082dec7acbaacc17f9   
76  d457916b4fdc60154ed93b5dd3e6fd69  76d64c4aca3a7baf218bf93ef7fa768d   
79  a659cb33082b851fb87a33af8f0fff29  817245bcc3badd82bbd222e0366951a6   

                           review_id  
14  15898b543726a832d4137fbef5d1d00e  
60  25b502beeef8b3aff87449ccdc001f53  
66  4ff7c437a

In [8]:
#verificación nombres categorias correctos
print("Categorias Order status:", df_u['order_status'].unique())
print("Categorias Payment type:", df_u['payment_type'].unique())
print("Categorias product_category_name english:", df_u['product_category_name_english'].unique())

Categorias Order status: ['delivered' 'invoiced' 'shipped' 'processing' 'unavailable' 'canceled'
 'created' 'approved']
Categorias Payment type: ['credit_card' 'voucher' 'boleto' 'debit_card' 'not_defined' nan]
Categorias product_category_name english: ['housewares' 'perfumery' 'auto' 'pet_shop' 'stationery' nan
 'furniture_decor' 'office_furniture' 'garden_tools'
 'computers_accessories' 'bed_bath_table' 'toys'
 'construction_tools_construction' 'telephony' 'health_beauty'
 'electronics' 'baby' 'cool_stuff' 'watches_gifts' 'air_conditioning'
 'sports_leisure' 'books_general_interest' 'small_appliances' 'food'
 'luggage_accessories' 'fashion_underwear_beach' 'christmas_supplies'
 'fashion_bags_accessories' 'musical_instruments'
 'construction_tools_lights' 'books_technical' 'costruction_tools_garden'
 'home_appliances' 'market_place' 'agro_industry_and_commerce'
 'party_supplies' 'home_confort' 'cds_dvds_musicals'
 'industry_commerce_and_business' 'consoles_games' 'furniture_bedroom'
 

In [9]:
#guardar data frame en nuevo csv olist_datos
df_u.to_csv("../datos/olist_datos.csv", index=False)


#### Calidad de Datos

**Valores nulos significativos** en columnas clave:
- **order_delivered_customer_date:** 2871 nulos (2.41%) - pedidos aún no entregados
- **review_score:** 99181 nulos (83.24%) - la mayoría de clientes no deja reseña
- **Dimensiones de productos:** 72% de nulos en peso, largo, alto, ancho

**Duplicados** Se detectaron 11303 filas duplicadas, esperadas por la estructura relacional (un pedido puede tener múltiples ítems o métodos de pago).

#### Distribuciones Clave

**Operaciones:** El 97% de pedidos están en estado "delivered", indicando operación logística muy exitosa.

**Pagos:** La tarjeta de crédito domina con 74% de transacciones, seguida de boleto bancario (19%). Esto refleja la preferencia brasileña por financiamiento.

**Geografía:** Fuerte concentración en São Paulo (42% clientes, 71% vendedores), mientras estados remotos tienen presencia mínima (<0.1%).

**Productos:** Catálogo diversificado con bed_bath_table liderando, pero ninguna categoría domina completamente (TOP categoría: 10%).

**Satisfacción:** Review score medio de **4.02/5 estrellas**, con distribución sesgada hacia valoraciones positivas (mayoría 4-5 estrellas).

#### Estadísticas Económicas

- Precio medio del producto: R$120.65 (productos accesibles)
- Costo medio de envío: R$20.03 (16% del precio)
- Ticket promedio total: R$172.74
- Productos principalmente ligeros: peso medio de 2112g